File: backend/app/schemas/activity.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


ResearchActivityType = Literal[
    "paper_opened",
    "paper_completed",
    "paper_saved",
    "paper_skipped",
    "paper_revisited",
]


class ResearchActivityCreate(BaseModel):
    paper_id: str = Field(min_length=1, max_length=255)
    event_type: ResearchActivityType


class ResearchActivityResponse(BaseModel):
    id: int
    paper_id: str
    event_type: ResearchActivityType
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/services/activity.ts
import api from "./api";

export type ResearchActivityType =
  | "paper_opened"
  | "paper_completed"
  | "paper_saved"
  | "paper_skipped"
  | "paper_revisited";

export async function recordResearchActivity(
  paperId: string,
  eventType: ResearchActivityType,
): Promise<void> {
  await api.post("/activity", {
    paper_id: paperId,
    event_type: eventType,
  });
}


export interface ResearchActivity {
  id: number;
  paper_id: string;
  event_type: ResearchActivityType;
  created_at: string;
}

export async function getRecentResearchActivity(): Promise<ResearchActivity[]> {
  const response = await api.get<ResearchActivity[]>("/activity");
  return response.data;
}
